### PR TITLE
feat!: add ability to use `react-native-test-app` for `example/` app

### PR DESCRIPTION
**Pull request by szymonrybczak**
### Summary

Use [`react-native-test-app`](https://github.com/microsoft/react-native-test-app/) for `exmaple/` app. RNTA was created to help library maintainers focus on building library by providing seamless React Native versions switching, first-class support for OOT platforms and a lot more! 

### Test plan

Generated app should work in the same way as before (created library should be automatically linked).

CleanShot 2024-06-27 at 16 51 30@2x

### CODE REVIEW

File: packages/create-react-native-library/src/utils/generateExampleApp.ts [hunk 1]
```diff
@@ -65,17 +65,22 @@ export default async function generateExampleApp({
...
-          reactNativeVersion,
-          '--skip-install',
-          '--npm',
+          ...(reactNativeVersion !== 'latest'
+            ? ['--version', reactNativeVersion]
+            : []),
+          '--platform',
+          'ios',
+          '--platform',
+          'android',
```
Comment (szymonrybczak) [inline packages/create-react-native-library/src/utils/generateExampleApp.ts:83]: Q: should we ask users which platforms they want to add to `example/` app?

File: packages/create-react-native-library/src/utils/generateExampleApp.ts [hunk 2]
```diff
@@ -65,17 +65,22 @@ export default async function generateExampleApp({
...
-          '--directory',
+          `--destination`,
           directory,
-          '--version',
-          reactNativeVersion,
-          '--skip-install',
-          '--npm',
+          ...(reactNativeVersion !== 'latest'
+            ? ['--version', reactNativeVersion]
+            : []),
```
Comment (szymonrybczak) [inline packages/create-react-native-library/src/utils/generateExampleApp.ts:79]: RNTA is failing when providing `latest` as value for `--version` option (it automatically fetches latest version).

File: packages/create-react-native-library/templates/native-common-example/example/react-native.config.js [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
 
 module.exports = {
+  project: configureProjects({
+    android: {
+      sourceDir: 'android',
+    },
+    ios: {
+      sourceDir: 'ios',
+    },
+  }),
```
[the remark this answers is not in the document]
Comment (szymonrybczak) [reply packages/create-react-native-library/templates/native-common-example/example/react-native.config.js:13]: now yes, agree.

File: packages/create-react-native-library/src/index.ts [hunk 4]
```diff
@@ -139,11 +139,17 @@ type Answers = {
...
+  example?: ExampleType;
   reactNativeVersion?: string;
   withRecommendedOptions?: boolean;
 };
 
+export enum ExampleType {
+  Vanilla = 'vanilla',
+  TestApp = 'test-app',
+  Expo = 'expo',
+}
```
[the remark this answers is not in the document]
Comment (szymonrybczak) [reply packages/create-react-native-library/src/index.ts]: https://github.com/callstack/react-native-builder-bob/pull/572/commits/6a2f3daeb8f9a4eee390136a13dd8175cab19ddf

File: packages/create-react-native-library/src/utils/generateExampleApp.ts (hunk 1 above)
Comment (szymonrybczak) [reply packages/create-react-native-library/src/utils/generateExampleApp.ts]: and also adding new platform for existing RNTA project is just a matter of [one command](https://github.com/microsoft/react-native-test-app/wiki/Updating#adding-platforms-to-an-existing-app).

File: packages/create-react-native-library/src/index.ts [hunk 5]
```diff
@@ -632,7 +635,7 @@ async function create(argv: yargs.Arguments<any>) {
     : 'legacy';
 
   const example =
-    hasExample && !local ? (type === 'library' ? 'expo' : 'native') : 'none';
+    hasExample && !local ? (type === 'library' ? 'expo' : hasExample) : null;
```
Comment (satya164) [inline packages/create-react-native-library/src/index.ts:638]: Does `test-app` have web support? Wondering if we can use it for `library` type

Comment (szymonrybczak) [reply packages/create-react-native-library/src/index.ts:638]: https://github.com/microsoft/react-native-test-app/issues/812#issuecomment-1071134434

File: packages/create-react-native-library/templates/native-common-example/example/react-native.config.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (szymonrybczak) [reply packages/create-react-native-library/templates/native-common-example/example/react-native.config.js]: https://github.com/callstack/react-native-builder-bob/pull/572/commits/5a70b276b6feca3734de8afce8dc6ed2cf83c366

**Review by satya164 — APPROVED** [review]
Thanks!

**Pull request merged**